### PR TITLE
feat(mempool): fix overflow bug in fee escalation flow; add tests

### PR DESCRIPTION
**Pull request by elintul**

### CODE REVIEW

**Comment by lotem-starkware** [comment]
This change is [Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/1447)

**Review by elintul — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/1447)* status: 0 of 3 files reviewed, 1 unresolved discussion (waiting on @ayeletstarkware, @giladchase, and @MohammadNassar1)

*[a discussion](https://reviewable.io/reviews/starkware-libs/sequencer/1447#-O9FVexN5-9WNGGgT_rP:-O9FVexN5-9WNGGgT_rQ:b-3hpxbx) (no related file):*
Depends on: #1445, for CI to pass.

___

**Review by MohammadNassar1 — COMMENTED** [review]
> Reviewed 1 of 3 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/1447)* status: 1 of 3 files reviewed, 1 unresolved discussion (waiting on @ayeletstarkware, @elintul, and @giladchase)

___
*[`crates/mempool/src/mempool.rs` line 251 at r2](https://reviewable.io/reviews/starkware-libs/sequencer/1447#-O9cqmMk7vuxuuNI2wv9:-O9cqmMk7vuxuuNI2wvA:b-hzqimt) ([raw file](https://github.com/starkware-libs/sequencer/blob/cbb685f794c64cad0496e7d5723c9afebe40a8df/crates/mempool/src/mempool.rs#L251)):*
> ```XML
>         let percentage = u128::from(self.fee_escalation_percentage);
> 
>         let Some(escalation_qualified_value) = (existing_value / 100)
> ```

Can you please document why it's okay to divide it before the multiplication?

Does it depend on the ranges of the gas/tip values?

What's the range of these values?



_Code quote:_
```XML
(existing_value / 100)
```

**Review by elintul — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/1447)* status: 1 of 3 files reviewed, 1 unresolved discussion (waiting on @ayeletstarkware, @giladchase, and @MohammadNassar1)

___
*[`crates/mempool/src/mempool.rs` line 251 at r2](https://reviewable.io/reviews/starkware-libs/sequencer/1447#-O9cqmMk7vuxuuNI2wv9:-O9d6I295-3vKDEjrlw8:b-u4v73a) ([raw file](https://github.com/starkware-libs/sequencer/blob/cbb685f794c64cad0496e7d5723c9afebe40a8df/crates/mempool/src/mempool.rs#L251)):*
<details><summary><i>Previously, MohammadNassar1 (mohammad-starkware) wrote…</i></summary><blockquote>

Can you please document why it's okay to divide it before the multiplication?

Does it depend on the ranges of the gas/tip values?

What's the range of these values?
</blockquote></details>

Why wouldn't it be okay? Value is `u128`, clear from the signature. Not sure what information I can add with words.

**Review by MohammadNassar1 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/1447)* status: 1 of 3 files reviewed, 1 unresolved discussion (waiting on @ayeletstarkware, @elintul, and @giladchase)

___
*[`crates/mempool/src/mempool.rs` line 251 at r2](https://reviewable.io/reviews/starkware-libs/sequencer/1447#-O9cqmMk7vuxuuNI2wv9:-O9dLq_u8In2hhI-8BUD:b-gv1u6l) ([raw file](https://github.com/starkware-libs/sequencer/blob/cbb685f794c64cad0496e7d5723c9afebe40a8df/crates/mempool/src/mempool.rs#L251)):*
<details><summary><i>Previously, elintul (Elin) wrote…</i></summary><blockquote>

Why wouldn't it be okay? Value is `u128`, clear from the signature. Not sure what information I can add with words.
</blockquote></details>

Typically, to avoid issues with integer division, we multiply first and then divide. However, in this case, we use a different approach.

This can cause problems when `existing_value`s are less than 100..
Do you assume that tip/gas are above these values? is it worth mentioning?

Another Q that I asked above: what's the range of these values?

**Review by elintul — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/1447)* status: 1 of 3 files reviewed, 1 unresolved discussion (waiting on @ayeletstarkware, @giladchase, and @MohammadNassar1)

___
*[`crates/mempool/src/mempool.rs` line 251 at r2](https://reviewable.io/reviews/starkware-libs/sequencer/1447#-O9cqmMk7vuxuuNI2wv9:-O9dWjo31kRWS66fhjEy:b-i6hfyd) ([raw file](https://github.com/starkware-libs/sequencer/blob/cbb685f794c64cad0496e7d5723c9afebe40a8df/crates/mempool/src/mempool.rs#L251)):*
<details><summary><i>Previously, MohammadNassar1 (mohammad-starkware) wrote…</i></summary><blockquote>

Typically, to avoid issues with integer division, we multiply first and then divide. However, in this case, we use a different approach.

This can cause problems when `existing_value`s are less than 100..
Do you assume that tip/gas are above these values? is it worth mentioning?

Another Q that I asked above: what's the range of these values?
</blockquote></details>

Okay good point, I'll give it some more thought; changed the order in the meantime in order not to delay this PR.
WDYM by range? Can you rephrase differently? I thought I answered: you know from the type.

**Review by MohammadNassar1 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/1447)* status: 1 of 3 files reviewed, 1 unresolved discussion (waiting on @ayeletstarkware, @elintul, and @giladchase)

___
*[`crates/mempool/src/mempool.rs` line 251 at r2](https://reviewable.io/reviews/starkware-libs/sequencer/1447#-O9cqmMk7vuxuuNI2wv9:-O9dpqTC4MvIy9ywtQgk:brvsv8g) ([raw file](https://github.com/starkware-libs/sequencer/blob/cbb685f794c64cad0496e7d5723c9afebe40a8df/crates/mempool/src/mempool.rs#L251)):*
<details><summary><i>Previously, elintul (Elin) wrote…</i></summary><blockquote>

Okay good point, I'll give it some more thought; changed the order in the meantime in order not to delay this PR.
WDYM by range? Can you rephrase differently? I thought I answered: you know from the type.
</blockquote></details>

Is it acceptable for the 'tip' or 'gas' values to be small (smaller than 100 in this case)? While the parameter type allows it, does the UX permit such small values?

**Review by MohammadNassar1 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/1447)* status: 1 of 3 files reviewed, 2 unresolved discussions (waiting on @ayeletstarkware, @elintul, and @giladchase)

___
*[`crates/mempool/src/mempool_test.rs` line 803 at r3](https://reviewable.io/reviews/starkware-libs/sequencer/1447#-O9dvaT4EVhFSj7IcMWt:-O9dvaT4EVhFSj7IcMWu:bs6x6wc) ([raw file](https://github.com/starkware-libs/sequencer/blob/577e96bada178518f57852ac124c25269faf65b2/crates/mempool/src/mempool_test.rs#L803)):*
> ```XML
>     for (tip, max_l2_gas_price) in initial_values {
>         let existing_tx = tx!(tip: tip, max_l2_gas_price: max_l2_gas_price, tx_nonce: 1,
>             sender_address: "0x0");
> ```

Why do you pass sender_address? I see you always use the default address.



_Code quote:_
```XML
sender_address: "0x0"
```

**Review by MohammadNassar1 — APPROVED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/1447#-:-O9dycoyAgrcbtbhCLZh:bnfp4nl)* status: 1 of 3 files reviewed, 1 unresolved discussion (waiting on @ayeletstarkware, @elintul, and @giladchase)

**Review by MohammadNassar1 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/1447)* status: 1 of 3 files reviewed, 2 unresolved discussions (waiting on @ayeletstarkware, @elintul, and @giladchase)

___
*[`crates/mempool/src/mempool.rs` line 252 at r3](https://reviewable.io/reviews/starkware-libs/sequencer/1447#-O9dyh__C77I-l4R3hOP:-O9dyh__C77I-l4R3hOQ:bimxcad) ([raw file](https://github.com/starkware-libs/sequencer/blob/edb7e0149ef8ce1ffeb65fe289b4105c8d2ae4d1/crates/mempool/src/mempool.rs#L252)):*
> ```XML
>         let escalation_factor = 100 + u128::from(self.fee_escalation_percentage);
> 
>         // TODO(Elin): add overflow tests; 2^127 existing with 10% increase should not overflow.
> ```

Do you want to keep the TODO?



_Code quote:_
```XML
// TODO(Elin): add overflow tests; 2^127 existing with 10% increase should not overflow.
```

**Review by elintul — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/1447)* status: 1 of 3 files reviewed, 2 unresolved discussions (waiting on @ayeletstarkware, @giladchase, and @MohammadNassar1)

___
*[`crates/mempool/src/mempool_test.rs` line 803 at r3](https://reviewable.io/reviews/starkware-libs/sequencer/1447#-O9dvaT4EVhFSj7IcMWt:-O9e0m5I9Xbk28FF8Fr0:b-b3d24d) ([raw file](https://github.com/starkware-libs/sequencer/blob/577e96bada178518f57852ac124c25269faf65b2/crates/mempool/src/mempool_test.rs#L803)):*
<details><summary><i>Previously, MohammadNassar1 (mohammad-starkware) wrote…</i></summary><blockquote>

Why do you pass sender_address? I see you always use the default address.
</blockquote></details>

Since it appears in the error message.

**Review by MohammadNassar1 — COMMENTED** [review]
> Reviewed 1 of 2 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/1447)* status: 2 of 3 files reviewed, 2 unresolved discussions (waiting on @ayeletstarkware, @elintul, and @giladchase)

___
*[`crates/mempool/src/mempool_test.rs` line 803 at r3](https://reviewable.io/reviews/starkware-libs/sequencer/1447#-O9dvaT4EVhFSj7IcMWt:-O9e8lBt-pAR286jcYir:b-ddo9r5) ([raw file](https://github.com/starkware-libs/sequencer/blob/577e96bada178518f57852ac124c25269faf65b2/crates/mempool/src/mempool_test.rs#L803)):*
<details><summary><i>Previously, elintul (Elin) wrote…</i></summary><blockquote>

Since it appears in the error message.
</blockquote></details>

Are you referring to the error in `add_txs_and_verify_no_replacement`? Since it's not part of the test, I believe you can use the default value. 

But, feel free to disregard this comment and go ahead with merging the PR.

**Review by elintul — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/1447)* status: 2 of 3 files reviewed, 2 unresolved discussions (waiting on @ayeletstarkware, @giladchase, and @MohammadNassar1)

___
*[`crates/mempool/src/mempool_test.rs` line 803 at r3](https://reviewable.io/reviews/starkware-libs/sequencer/1447#-O9dvaT4EVhFSj7IcMWt:-O9eBOxC7ZJKi2ftEVdw:bnywjom) ([raw file](https://github.com/starkware-libs/sequencer/blob/577e96bada178518f57852ac124c25269faf65b2/crates/mempool/src/mempool_test.rs#L803)):*
<details><summary><i>Previously, MohammadNassar1 (mohammad-starkware) wrote…</i></summary><blockquote>

Are you referring to the error in `add_txs_and_verify_no_replacement`? Since it's not part of the test, I believe you can use the default value. 

But, feel free to disregard this comment and go ahead with merging the PR.
</blockquote></details>

Yeah, perhaps it isn't needed here, since the address is now inferred from the existing transaction.
Good catch, I'll go over all fee escalation tests and fix it.

**Review by MohammadNassar1 — APPROVED** [review]
> Reviewed 1 of 2 files at r3.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/1447)* status: all files reviewed, 1 unresolved discussion (waiting on @ayeletstarkware, @elintul, and @giladchase)

**Review by elintul — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/1447)* status: 2 of 3 files reviewed, 1 unresolved discussion (waiting on @ayeletstarkware, @giladchase, and @MohammadNassar1)

___
*[`crates/mempool/src/mempool.rs` line 252 at r3](https://reviewable.io/reviews/starkware-libs/sequencer/1447#-O9dyh__C77I-l4R3hOP:-O9hR8Tj1gj53hOseDSp:b8qk2hf) ([raw file](https://github.com/starkware-libs/sequencer/blob/edb7e0149ef8ce1ffeb65fe289b4105c8d2ae4d1/crates/mempool/src/mempool.rs#L252)):*
<details><summary><i>Previously, MohammadNassar1 (mohammad-starkware) wrote…</i></summary><blockquote>

Do you want to keep the TODO?
</blockquote></details>

No, I have the comment above the test.

**Review by MohammadNassar1 — APPROVED** [review]
> Reviewed 2 of 2 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/1447)* status: <span>:shipit:</span> complete! all files reviewed, all discussions resolved (waiting on @ayeletstarkware and @giladchase)
